### PR TITLE
DataLinks: Fix bug where links which used built in variables could be hidden

### LINKED ISSUE #70809: Tempo Service Graph: Explore links to prometheus are not rendered
### 
# What went wrong?

**What happened**:

Context menu for node panel nodes & edges only has explore link to traces.

**What did you expect to happen**:

Context menu for node panel nodes & edges to also have links to explore as defiend here https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/tempo/datasource.ts#L695

### 
# How do we reproduce it?

Open any node graph, click on a node, check that no prometheus links are rendered

### 
# What Grafana version are you using?

10.1

### 
### Is the bug inside a Dashboard Panel?

Copy the panel's ["get-help" data](https://grafana.com/docs/grafana/latest/troubleshooting/send-panel-to-grafana-support/) here

### Grafana Platform?

None

### User's OS?

_No response_

### User's Browser?

_No response_

### Is this a Regression?

None

### Are Datasources involved?

_No response_

### Anything else to add?

The root cause seems to be that prometheus links include `$__rate_interval`. This is treated as missing variable by [explore links util](https://github.com/grafana/grafana/blob/main/public/app/features/explore/utils/links.ts#L167) and links get omitted.

### PR DESCRIPTION
**Pull request by aocenas**
Fixes: https://github.com/grafana/grafana/issues/70809
Relates to https://github.com/grafana/grafana/issues/64328

At some point we added check to see if variables get interpolated in the link, mainly to not show link which would end up being broken because of missing values, this check though didn't consider built in variables.

This adds a static list of variables which we just ignore.

This is not 100% perfect, variables can be added later (for example in query time) and this needs to be synchronized if some variable gets added but should be good enough for now.

You can test it for example with Tempo service map, which when clicking on node should show multiple metrics links to Prometheus but all of them use `$__rate_interval` variable.

### CODE REVIEW

**Comment by gelicia** [comment]
> You can test it for example with Tempo service map

Are there instructions on how to get this set up with devenv tempo?

**Comment by gelicia** [comment]
Ah okay, so after taking a look at this and chatting with Andrej, the issue is interpolation can happen in two places (thank you for breaking it down for me!)

1. when creating the link, there we interpolate variables with data from the source
2. interpolation when query is executed, for example, prometheus does this [here](https://github.com/grafana/grafana/blob/72581a77acfa9fda2d5b45cb016d4c53afce3189/public/app/plugins/datasource/prometheus/datasource.tsx#L682) 

When building correlations, we added in logic under the assumption we would never want to show a query with an uninterpolated variable in it, so we hide all queries that still have variable values in them after that first interpolation. However, queries can pass through with an uninterpolated variable, and then be interpolated later by the datasource code. Those datasources have a set list of expected variables they'd interpolate, so this PR adds them so the correlations code doesn't stop them from showing.

I'm super torn about this, I agreed having the uninterpolated variables displaying didn't look right, but having to maintain this list and having exceptions is also not great. If we do go this route we should have the list of exceptions in the correlations documentation.

**Comment by ifrost** [comment]
Good spot!

As it's a regression, it should be okay for a short-term quick fix but it seems we need a better solution. `$__rate_interval` is not a built-in variable, I think it's Prometheus-specific, it won't scale work long term. A few options we could consider to fix it:

* Add `conditions` to data links config (correlations would set `requiresAllVariables` but I can imagine this could be expanded in the future with more conditions). We kind of tried that in the pass but it was hard to implement just for Explore. Data sources would not set such conditions (but can do it optionally)
* Add escaping mechanism to templating so a variable can actually be included inside the query (something like https://github.com/grafana/grafana/pull/71424 ?)
* Use DataSourceAPI.interpolateVariablesInQueries to interpolate the link (though I think Prometheus is not actually using it there)

Two more things about this PR:
- It's a regression, so we should add a test for it
- Do we need to backport it?

File: public/app/features/explore/utils/links.test.ts [hunk 1]
```diff
@@ -644,86 +644,47 @@ describe('explore links utils', () => {
...
-          datasourceUid: 'uid',
-          datasourceName: 'dsName',
-          query: { query: 'test ${diffVar}' },
-        },
-      };
-      const scopedVars = {
-        testVal: { text: '', value: 'val1' },
-      };
-      const dataLinkRtnVal = getVariableUsageInfo(dataLink, scopedVars).allVariablesDefined;
+    it('ignores global variables', () => {
```
Comment (aocenas) [inline public/app/features/explore/utils/links.test.ts:670]: Only new test, others were just refactored to remove duplicated lines

**Review by gelicia — APPROVED** [review]

**Comment by gelicia** [comment]
Opened a ticket to add correlation documentation around this exception (#71511), and a ticket to develop a more robust solution for this scenario (#71510)

**Pull request merged**